[Mistral&Mixtral]Add sliding window param to sdpa after torch 2.2.0

ehuaa opened the pull request:
# What does this PR do?

Add sliding window param as described in https://github.com/huggingface/transformers/issues/28980, and the solution is check if torch version greater than or equal as torch 2.2.0 (release version) , and add sliding window  to attention_mask if it's true.
Tests of sdpa and sliding window attention with torch 2.2.0 has been passed with 
RUN_SLOW=yes python -m pytest -n auto --dist=loadfile -s -v ./tests/models/mistral/test_modeling_mistral.py
in my local environment.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ √] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ √] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ √] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker @fxmarty

ehuaa committed:
add sliding window param to sdpa after torch==2.2.0

ehuaa committed:
add sliding window param to sdpa after torch==2.2.0

ehuaa committed:
add sliding window param to sdpa after torch==2.2.0

ehuaa committed:
add sliding window param to sdpa after torch==2.2.0

ehuaa commented:
I modified mixtral and qwen2 model as well, because of they have same sliding window structure here.

ehuaa committed:
revert add sliding window for qwen2 because of numerical error

ehuaa committed:
revert add sliding window for qwen2 because of numerical error

ehuaa committed:
Merge branch 'add_sliding_window_for_sdpa' of https://mirror.ghproxy.com/https://github.com/ehuaa/transformers into add_sliding_window_for_sdpa

ehuaa committed:
remove adding sliding_window param to qwen2 because of numerical error

ehuaa committed:
remove adding sliding_window param to qwen2 because of numerical error

ehuaa committed:
fix style

ehuaa committed:
fix style

ehuaa committed:
only add non-contigous mask to qwen2 due to numerical error

ehuaa committed:
only add non-contigous mask to qwen2 due to numerical error

ArthurZucker commented on src/transformers/models/mistral/modeling_mistral.py:
```diff
@@ -680,7 +683,11 @@ def forward(
+            and query_states.device.type == "cuda"
+            and attention_mask is not None
+        ):
             query_states = query_states.contiguous()
             key_states = key_states.contiguous()
             value_states = value_states.contiguous()
```
If we are using torch2.2 this should not be super costly even if not necessary no?

ehuaa replied:
> If we are using torch2.2 this should not be super costly even if not necessary no?

Yes, you are right, it seems that not cost too much. I'll remove these three changes of making tensors non-contigous in torch 2.2.

ArthurZucker reviewed (commented):
Looks alright but let's not add the unrelated change

ehuaa commented on src/transformers/models/qwen2/modeling_qwen2.py:
```diff
@@ -690,7 +693,11 @@ def forward(
+        ):
             query_states = query_states.contiguous()
             key_states = key_states.contiguous()
             value_states = value_states.contiguous()
```
> missing sliding window here no?

when i add sliding window to qwen2, several tests of test_modeling_qwen2 goes wrong as i listed above, so i'll leave modeling_qwen2 unchanged.

ehuaa committed:
revert non-contigous tensor modification

ehuaa committed:
revert non-contigous tensor modification

ehuaa commented:
> Looks alright but let's not add the unrelated change

I have deleted the unrelated changes and please review them again. The failed test are because of 
error running git clone "git@github.com:huggingface/transformers.git": exit status 128, which is not related to this PR.

ArthurZucker commented on src/transformers/models/mistral/modeling_mistral.py:
```diff
@@ -51,6 +53,7 @@
 
     _flash_supports_window_size = "window_size" in list(inspect.signature(flash_attn_func).parameters)
 
+_is_torch_version_greater_or_equal_than_2_2_0 = version.parse(get_torch_version()) >= version.parse("2.2.0")
```
Let's do something close to 
https://github.com/huggingface/transformers/blob/main/src/transformers/utils/import_utils.py#L661

to make sure it's re-usable!

ehuaa replied:
> Let's do something close to https://github.com/huggingface/transformers/blob/main/src/transformers/utils/import_utils.py#L661
> 
> to make sure it's re-usable!

you're right, and i've added this function to import_utils for convenience， please review it again @ArthurZucker

ArthurZucker reviewed (commented):
A small nit but LGTM otherwise. 
Let's just declare `_is_torch_version_greater_or_equal_than_2_2_0` in the `src/transformers/utils/import_utils.py` file

ehuaa committed:
move torch version judgement to import_utils for usability

ehuaa committed:
move torch version judgement to import_utils for usability

ArthurZucker commented on src/transformers/models/mistral/modeling_mistral.py:
```diff
@@ -1006,6 +1007,7 @@ def forward(
                 (batch_size, seq_length),
                 inputs_embeds,
                 past_key_values_length,
+                sliding_window=self.config.sliding_window if is_torch_version_greater_or_equal_than_2_2_0 else None,
```
if the argument does not exist in previous versions, this will fail no? We should have a `kwarg` instead no?

ehuaa replied:
> if the argument does not exist in previous versions, this will fail no? We should have a `kwarg` instead no?

in the function definition listed below,
https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_attn_mask_utils.py#L336, there is a default None here, so i think this will not fail here in previous versions. @ArthurZucker

ArthurZucker replied:
Ah sorry sorry you are completely right!

ArthurZucker reviewed (commented):
Almost good to go

ArthurZucker reviewed (approved):
LGTM, let's fix the conflicts and we should be able to merge. 
EDIT: let's add a small test in the integration tests!

ehuaa commented:
> LGTM, let's fix the conflicts and we should be able to merge. EDIT: let's add a small test in the integration tests!

I have resolved the conflict, and the test you mentioned above i think we can use the test you wrote before, https://github.com/huggingface/transformers/blob/main/tests/models/mistral/test_modeling_mistral.py#L534
when you upgrade your torch version of huggingface environment to 2.2, and enable RUN_SLOW=yes, this test can check if this PR is right!
Additionaly, when we convert the tensor to contigous, it can solve the problem even with torch 2.1.1, so I think the test you wrote above is enough for testing!

ehuaa committed:
Merge branch 'main' into add_sliding_window_for_sdpa

ehuaa committed:
delete deprecated is_flash_attn in import_utils.py

ehuaa commented:
> When i tried to add a new test to check if sdpa work with sliding_window, the test below is failed.
> 
> I just modified your original test_model_7b_logits with a longer input with size of 4097 to check sliding_window feature, it failes as below: E AssertionError: Tensor-likes are not close! E E Mismatched elements: 1 / 4097 (0.0%) E Greatest absolute difference: 0.20025348663330078 at index (0, 4096) (up to 0.01 allowed) E Greatest relative difference: 0.06286640465259552 at index (0, 4096) (up to 0.01 allowed) which i think is also a problem with older versions of transformers. @ArthurZucker

@slow
    def test_model_7b_logits(self):
        input_ids = [1] + [306, 338] * 2048
        # input_ids = [1, 306, 4658, 278, 6593, 310, 2834, 338]
        with torch.backends.cuda.sdp_kernel(
                                        enable_flash=False,
                                        enable_math=True,
                                        enable_mem_efficient=False,
                                    ):
            model = MistralForCausalLM.from_pretrained("mistralai/Mistral-7B-v0.1", device_map="auto",attn_implementation="eager")
            input_ids = torch.tensor([input_ids]).to(model.model.embed_tokens.weight.device)
            with torch.no_grad():
                out = model(input_ids).logits.cpu()
        
        input_ids = [1] + [306, 338] * 2048  
        model = MistralForCausalLM.from_pretrained("mistralai/Mistral-7B-v0.1", device_map="auto",attn_implementation="sdpa")
        input_ids = torch.tensor([input_ids]).to(model.model.embed_tokens.weight.device)
        with torch.no_grad():
            out1 = model(input_ids).logits.cpu()
        # Expected mean on dim = -1
        torch.testing.assert_close(out.mean(-1), out1.mean(-1), atol=1e-2, rtol=1e-2)

        del model
        backend_empty_cache(torch_device)
        gc.collect()

SunMarc committed:
Add compatibility with skip_memory_metrics for mps device (#29264)

* Add compatibility with mps device

* fix

* typo and style

zucchini-nlp committed:
Token level timestamps for long-form generation in Whisper (#29148)

sadra-barikbin committed:
Fix a few typos in `GenerationMixin`'s docstring (#29277)

Co-authored-by: Joao Gante <joao@huggingface.co>

windsonsea committed:
[i18n-zh] Translate fsdp.md into Chinese (#29305)

* [i18n-zh] Translate fsdp.md into Chinese

Signed-off-by: windsonsea <haifeng.yao@daocloud.io>

* apply suggestions from Fan-Lin

---------

Signed-off-by: windsonsea <haifeng.yao@daocloud.io>

RaymondLi0 committed:
Starcoder2 model - bis (#29215)

* Copy model

* changes

* misc

* fixes

* add embed and residual dropout (#30)

* misc

* remove rms norm and gated MLP

* remove copied mentions where its not a copy anymore

* remove unused _shape

* copied from mistral instead

* fix copies

* fix copies

* add not doctested

* fix

* fix copyright

* Update docs/source/en/model_doc/starcoder2.md

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* Update src/transformers/models/starcoder2/configuration_starcoder2.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* Update src/transformers/models/starcoder2/configuration_starcoder2.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* fix doc

* revert some changes

* add fa2 tests

* fix styling nit

* fix

* push dummy docs

---------

Co-authored-by: Joel Lamy-Poirier <joel.lamy-poirier@servicenow.com>
Co-authored-by: younesbelkada <younesbelkada@gmail.com>
Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>
Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

cebtenzzre committed:
simplify get_class_in_module and fix for paths containing a dot (#29262)

younesbelkada committed:
FIX [`Gemma` / `CI`] Make sure our runners have access to the model (#29242)

* pu hf token in gemma tests

* update suggestion

* add to flax

* revert

* fix

* fixup

* forward contrib credits from discussion

---------

Co-authored-by: ArthurZucker <ArthurZucker@users.noreply.github.com>

fxmarty committed:
Remove numpy usage from owlvit (#29326)

* remove numpy usage from owlvit

* fix init owlv2

* style

ArthurZucker committed:
[`require_read_token`] fix typo (#29345)

fix wrapper

ArthurZucker committed:
[`T5 and Llama Tokenizer`] remove warning (#29346)

* remove warning

* add co-author

* update

---------

Co-authored-by: hiaoxui <hiaoxui@users.noreply.github.com>

ArthurZucker committed:
[`Llama ROPE`] Fix torch export but also slow downs in forward (#29198)

* remove control flow

* update gptneox

* update ....

* nits

* Actually let's just break. Otherwise we are silently failing which imo is not optimal

* version BC

* fix tests

* fix eager causal

* nit

* add a test

* style

* nits

* nits

* more nits for the test

* update and fix

* make sure cuda graphs are not skipped

* read token is needed for meta llama

* update!

* fiixup

* compile test should be slow

* fix thet fix copies

* stle 🫠

LeonardoEmili committed:
Disable Mixtral `output_router_logits` during inference (#29249)

* Set output_router_logits=False in prepare_inputs_for_generation for mixtral

* Add output_router_logits=False to prepare_inputs_for_generation for mixtral

* Fix style

gante committed:
Idefics: generate fix (#29320)

danielhanchen committed:
RoPE loses precision for Llama / Gemma + Gemma logits.float() (#29285)

* Update modeling_llama.py

Llama - Force float32 since bfloat16 loses precision on long contexts

* Update modeling_llama.py

* Update modeling_gemma.py

Fix RoPE and logits.float()

* @torch.no_grad()

* @torch.no_grad()

* Cos, Sin to float32

* cos, sin to float32

* Update src/transformers/models/gemma/modeling_gemma.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* Update src/transformers/models/llama/modeling_llama.py

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* Resolve PR conflicts

* Fix RoPE for llama

* Revert "Fix RoPE for llama"

This reverts commit b860a22dab9bb01cd15cb9a3220abeaefad3e458.

* Fix RoPE for llama

* RoPE device

* Autocast device type

* RoPE

* RoPE isinstance

---------

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

jiqing-feng committed:
check if position_ids exists before using it (#29306)

Co-authored-by: Joao Gante <joao@huggingface.co>

SunMarc committed:
[CI] Quantization workflow (#29046)

* [CI] Quantization workflow

* build dockerfile

* fix dockerfile

* update self-cheduled.yml

* test build dockerfile on push

* fix torch install

* udapte to python 3.10

* update aqlm version

* uncomment build dockerfile

* tests if the scheduler works

* fix docker

* do not trigger on psuh again

* add additional runs

* test again

* all good

* style

* Update .github/workflows/self-scheduled.yml

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

* test build dockerfile with torch 2.2.0

* fix extra

* clean

* revert changes

* Revert "revert changes"

This reverts commit 4cb52b8822da9d1786a821a33e867e4fcc00d8fd.

* revert correct change

---------

Co-authored-by: Younes Belkada <49240599+younesbelkada@users.noreply.github.com>

fxmarty committed:
Better SDPA unmasking implementation (#29318)

* better unmask imple

* comment

* typo

* bug report pytorch

* cleanup

* fix import

* add back example

* retrigger ci

* come on

windsonsea committed:
[i18n-zh] Sync source/zh/index.md (#29331)

* [i18n-zh] Sync source/zh/index.md

* apply review comments

younesbelkada committed:
FIX [`CI` / `starcoder2`] Change starcoder2 path to correct one for slow tests (#29359)

change starcoder2 path to correct one

younesbelkada committed:
FIX [`CI`]: Fix failing tests for peft integration (#29330)

fix failing tests for peft integration

younesbelkada committed:
FIX [`CI`] `require_read_token` in the llama FA2 test (#29361)

Update test_modeling_llama.py

ydshieh committed:
Avoid using uncessary `get_values(MODEL_MAPPING)` (#29362)

* more fixes

* more fixes

---------

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

NielsRogge committed:
Patch YOLOS and others (#29353)

Fix issue

Wauplin committed:
Fix @require_read_token in tests (#29367)

notsyncing committed:
Expose `offload_buffers` parameter of `accelerate` to `PreTrainedModel.from_pretrained` method (#28755)

Expose offload_buffers parameter to from_pretrained method

lenglaender committed:
Fix Base Model Name of LlamaForQuestionAnswering (#29258)

* LlamaForQuestionAnswering self.transformer->self.model

* fix "Copied from" string

* Llama QA model: set base_model_prefix = "transformer"

younesbelkada committed:
FIX [`quantization` / `ESM`] Fix ESM 8bit / 4bit with bitsandbytes (#29329)

* fix ESM 8bit

* Apply suggestions from code review

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* fixup

---------

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

ArthurZucker committed:
[`Llama + AWQ`] fix `prepare_inputs_for_generation`  🫠 (#29381)

* use the generation config 🫠

* fixup

sanchit-gandhi committed:
🚨🚨[Whisper Tok] Update integration test (#29368)

* [Whisper Tok] Update integration test

* make style

amyeroberts committed:
[`YOLOS`] Fix - return padded annotations (#29300)

* Fix yolos processing

* Add back slow marker - protects for pycocotools in slow

* Slow decorator goes above copied from header

JingyaHuang committed:
Support subfolder with `AutoProcessor` (#29169)

enable subfolder

psinger commented:
Thanks for this PR.
A quick related question that is confusing me a bit: Doesn't that mean that Mistral and related models were fully buggy the last few versions, as `sdpa` is the default config, but it does not respect `window_size`?

SunMarc committed:
Fix llama + gemma accelete tests (#29380)

muellerzr committed:
Fix deprecated arg issue (#29372)

* Fix deprecated arg issue

* Trainer check too

* Check for dict or dataclass

* Simplify, make config always AcceleratorConfig

* Upstream to Trainer

DavidAfonsoValente committed:
Correct zero division error in inverse sqrt scheduler (#28982)

* Correct zero division error in inverse sqrt scheduler

* default timescale to 10_000

ArthurZucker commented:
Probably yes. That why it's a bit critical. Mistral only had `flash_attention_2` at the beginning so I suspect people used this config explicitly. But after merging `sdpa` without sliding window if it was not explicitly specified yes. 
I don't really recalling having this reported 😢

ArthurZucker reviewed (commented):
The only thing missing for me is a test to make sure that `sdpa` with `sliding_window` gives the same results as `flash_attention_2` with `sliding_window`

ehuaa commented:
> The only thing missing for me is a test to make sure that `sdpa` with `sliding_window` gives the same results as `flash_attention_2` with `sliding_window`

@ArthurZucker , Yes, so i test mistral-7b logits of sdpa with sliding_window and eager with sliding_window, and the results are not the same.
The reason why not using flash_attention_2 is i only have a gpu card of v100 which is not compatiable with flash_attentnion_2, can you test it in your own machine above or should i use huggingface CLI environment to test?

ArthurZucker commented:
> Yes, so i test mistral-7b logits of sdpa with sliding_window and eager with sliding_window, and the results are not the same.
That's less important than sdpa vs flash! 
You can add the slow test already in `MistralIntegrationTest` and I will run it on my machines

ehuaa commented:
> > Yes, so i test mistral-7b logits of sdpa with sliding_window and eager with sliding_window, and the results are not the same.
> 
> That's less important than sdpa vs flash! You can add the slow test already in `MistralIntegrationTest` and I will run it on my machines
Ok, I'll upload a new slow test of sdpa vs flash later.

ehuaa committed:
upload a test for compare flash vs sdpa for sliding window in Mistral

ehuaa committed:
Merge branch 'add_sliding_window_for_sdpa' of https://mirror.ghproxy.com/https://github.com/ehuaa/transformers into add_sliding_window_for_sdpa

ehuaa commented:
After I git push, there're something weird to the changed files above so i closed this pr and try to open a new pull request. @ArthurZucker

ArthurZucker commented:
alright! It think this fix is important, let's try to have this ready! Ping me on the next PR and link it with this one as well to get the full conv!

ehuaa commented:
> alright! It think this fix is important, let's try to have this ready! Ping me on the next PR and link it with this one as well to get the full conv!

https://github.com/huggingface/transformers/pull/29407 @ArthurZucker  Please review this pr you mentioned above.
